File: hardcore-sys/hardcore/src/api/render/ops/raster_commands.cpp

#include <pch.hpp>

#include <render/ops/raster_commands.h>

#include "../../validation.hpp"

#include <render/renderer.hpp>

#include <core/error.hpp>
#include <core/log.hpp>
#include <util/number.hpp>

HCResult hc_new_draw(
    HCDraw* draw,
    HCRenderPass const* render_pass,
    u32 subpass,
    HCRasterPipeline const* pipeline,
    u32 vertex_count,
    u32 instance_count
) {
    HC_VALIDATE_PTR_RE(draw, "draw");
    HC_VALIDATE_PTR_RE(render_pass, "render pass");
    HC_VALIDATE_PTR_RE(pipeline, "raster pipeline");
    HC_VALIDATE_RE(render_pass->device == pipeline->device, "Render pass and pipeline devices do not match");

    auto device_result = hc::render::device_at(render_pass->device);
    if (!device_result) {
        return device_result.error();
    }

    auto const result = device_result.value()->create_draw(render_pass->id, subpass, pipeline->id, vertex_count, instance_count);
    if (!result) {
        return result.error();
    }

    *draw = {
        .id = *result,
        .device = render_pass->device,
    };

    return { .success = true };
}

void hc_destroy_draw(HCDraw* draw) {
    HC_VALIDATE_PTR(draw, "draw");

    auto device_result = hc::render::device_at(draw->device);
    if (!device_result) {
        return;
    }

    device_result.value()->destroy_draw(draw->id);

    *draw = {};
}

HCResult hc_set_draw_push_constants(HCDraw const* draw, void const** push_constants, u32 count) {
    HC_VALIDATE_PTR_RE(draw, "draw");
    HC_VALIDATE_RE(count > 0, "Invalid push constant count");
    HC_VALIDATE_PTR_RANGE_RE(push_constants, count, "push constants");

    auto device_result = hc::render::device_at(draw->device);
    if (!device_result) {
        return device_result.error();
    }

    auto result = device_result.value()->set_draw_push_constants(draw->id, std::span(push_constants, count));
    if (!result) {
        return result.error();
    }

    return { .success = true };
}

File: hardcore/src/ops/raster_commands.rs
use crate::handle::Handle;
use crate::ops::raster_pipeline::RasterPipeline;
use crate::ops::render_pass::RenderPass;
use crate::Error;

pub struct Draw<'s> {
    handle: Handle<'s, hardcore_sys::Draw>,
}

impl<'s> Draw<'s> {
    pub(crate) fn new(
        render_pass: &RenderPass<'s>,
        subpass: u32,
        pipeline: &RasterPipeline<'s>,
        vertex_count: u32,
        instance_count: u32,
    ) -> Result<Draw<'s>, Error> {
        if render_pass.handle.inner.device != pipeline.handle.inner.device {
            return Err(Error::InvalidParams(
                "Render pass and pipeline do no belong to the same device".to_string(),
            ));
        }

        let mut handle = Default::default();

        unsafe {
            hardcore_sys::new_draw(
                &raw mut handle,
                render_pass.handle.ptr(),
                subpass,
                pipeline.handle.ptr(),
                vertex_count,
                instance_count,
            )
            .into_std_result()?;
        }

        Ok(Draw {
            handle: handle.into(),
        })
    }

    // pub fn set_push_constants(&self) -> Result<(), Error> {
    //     unsafe { hardcore_sys::set_draw_push_constants(self.handle.ptr()) }
    // }
}

impl Drop for Draw<'_> {
    fn drop(&mut self) {
        unsafe { hardcore_sys::destroy_draw(self.handle.mut_ptr()) }
    }
}
